Documentation/PhpArchitecture/Readonly.rst

```diff
@@ -0,0 +1,131 @@
+influencing possible further usages within the same request. Statelessness
+is an important asset of services and readonly helps to sort out this question.
+
+Even when a service class is declared as readonly, ensuring immutability at its level,
+it can still become stateful if any of its injected dependencies are stateful. This
+undermines the benefits of readonly design, as statefulness in dependencies can
+introduce unintended side effects and compromise the stateless nature of the service.
+TYPO3 Core strives to avoid such scenarios, particularly for services that are widely
+used by extensions. This ensures predictable behavior, minimizes side effects, and
+maintains consistency in the broader ecosystem. Developers should carefully analyze
+dependencies for statefulness when designing readonly services.
+
+The TYPO3 Core development adopted the readonly feature early, recognizing its
+advantages for improving immutability, reducing side effects, and clarifying
+service design. However, its use requires careful consideration. The Core merger
+team established guidelines to determine when readonly can and should be added,
+which also serve as best practices for extension developers:
+
+*   **General Recommendation:** Declaring services or their properties as readonly
+    is highly encouraged. Once added, the readonly declaration is rarely removed since
+    it aligns with the effort to make services stateless.
+
+*   **Leaf Classes:** Existing services that are "leaf" classes (i.e., not intended
+    to be extended by other classes) can have readonly applied to single properties
+    or the entire class. This is typically not considered a breaking change, even in
+    stable code branches, as it only affects XCLASS extensions, which are not covered
+    by TYPO3's backward compatibility promise.
+
+*   **Method Injection:** Services retrieved via :ref:`inject*() methods <method-injection>`
+    are not currently declared readonly, as tools like PHPStan expect readonly properties
+    to be initialized in the constructor. This might change in the future, but it is
+    not a high priority.
+
+*   **Abstract Classes:** Existing abstract classes that are intended for extension by
+    developers should not be declared readonly. Declaring an abstract class readonly
+    would force all inheriting classes to also be readonly, which can create compatibility
+    issues for extensions that need to support multiple TYPO3 versions. For example,
+    Extbase's ActionController will not be declared readonly.
+
+
+Readonly value objects
+----------------------
+
+Readonly value objects are immutable by design. They align seamlessly with
+public constructor property promotion for simplicity:
+
+..  code-block:: php
+    :caption: Read only value object using public constructor property promotion
+
+    final readonly class Label
+    {
+        public function __construct(
+            public string $label,
```

Comment:
You told me that building such value objects that way means these property names are API now. You have to discuss about these names very clearly. Changing a name would break "API". Do not add something like `$containsNameOrTitleOfRecord`. Be very precise here. I also would try to prevent the use of `mixed` type or multiple types each property here.

Reply:
discussing internals of final readonly data objects is not scope of this readonly document. coreapi already has some places to talk about data objects, this could be streamlined a bit with further patches. i'll not add extensive information here and hope that is ok with you.